### PR TITLE
Only set has changed for field if value changed

### PR DESCRIPTION
**Pull request by ianworth**
**Summary**

We've encountered a similar issue to that reported in #725 where the inclusion of a relation widget in the editor page causes an article to require hitting the save button twice to save the article.  This seems to be because the setting of metadata in the a DRAFT_CHANGE_FIELD signal is changing hasChanged attribute in the entry state to be true even though the value has not changed at all.

I'm new to the internals of Netlify CMS and various logic flows, but I'm wondering whether the metadata changing should really be indicating that entry has changed, or whether this meta data is just used for the CMS interface.  Any guidance on any side effects of this PR would be appreciated to help me steer to the correct solution.

**Test plan**

Unit tests added.

Functional testing, before fix 

Given GitHub backend & editorial workflow enabled (in case that makes a difference) 
And article with Relation widget in place
When I create new article
And click save
Then article saves, but the "unsaved changes" message appears and I need to save the changes

After this fix in this PR

When I create new article
And click save
Then article saves and the message remains as "changes saved" allowing me to continue with the workflow.

**A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by ianhomer** [comment]
BTW - I can rebase and squash if I'm on the right track ...

**Comment by erquhart** [comment]
Thanks for this! I'm not totally certain about side effects - the associated code paths need some attention. I hesitate on this PR due to the unknowns and also because the main occurrences of this issue are still present in the deploy preview. The general test is to open a new post in the demo and see if there are unsaved changes: https://deploy-preview-2377--cms-demo.netlify.com/#/collections/posts/new

**Comment by barthc** [comment]
> but I'm wondering whether the metadata changing should really be indicating that entry has changed, or whether this meta data is just used for the CMS interface.

The relation widget `metadata` is only used for the editor preview AFIAK. The issue this PR is trying to address is also caused by the `date` widget that is why we can still the issue here  https://deploy-preview-2377--cms-demo.netlify.com/#/collections/posts/new

**Comment by ianworth** [comment]
thanks for the feedback - you're right, this PR is actually fixing a different issue to that reported in #725 - I'll have a look at the cause of #725 and see if I can fix that too.

**Comment by ianhomer** [comment]
Not entirely sure this is the right way to crack this ... however I've allowed a widget control to define a static method called "createDefaultValue" that is loaded as a defaultValue function as the config is loaded.  Then when we createEmptyDraftData I can call this function to get the default value.

Initial quick testing seems to fix #725 in that when I create a new post with a date field then 1) the current date set as the initial value for the date 2) "unsaved changes" is only set when I make an explicit change 3) I can click away without any pop up if I've made no changes.

**Comment by ianhomer** [comment]
Note I've updated the test content, since a few (required) date / datetime fields were empty in the test content which was leading to react date control setting them to moment() on load and triggering a content change signal on loading that article in the CMS editor.

**Comment by ianhomer** [comment]
Preview at https://deploy-preview-2377--cms-demo.netlify.com now allows us to load the following without flipping the unsaved changes flag

* [new post](https://deploy-preview-2377--cms-demo.netlify.com/#/collections/posts/new)
* [new kitchen sink](https://deploy-preview-2377--cms-demo.netlify.com/#/collections/kitchenSink/new)
* [edit existing kitchen sink](https://deploy-preview-2377--cms-demo.netlify.com/#/collections/kitchenSink/entries/a-big-entry-with-all-the-things)

**Comment by erquhart** [comment]
@barthc can you give this a quick look? Might conflict with or impact your recent work on default values.

File: packages/netlify-cms-core/src/actions/config.js [hunk 1]
```diff
@@ -61,9 +62,37 @@ export function applyDefaults(config) {
...
+function bindCreateDefaultValues(fields) {
+  return fields.map(field => {
+    return field.withMutations(field => {
+      if (!field.get('default')) {
+        const createDefaultValue = resolveWidget(field.get('widget')).control.createDefaultValue;
+        if (createDefaultValue) {
+          field.set('default', createDefaultValue);
+        }
+      }
+      if (field.get('fields')) {
```
Comment (barthc) [inline packages/netlify-cms-core/src/actions/config.js]: @ianhomer I think you also need to account for `field` widgets

Comment (erquhart) [reply packages/netlify-cms-core/src/actions/config.js]: Yeah I was thinking the same - probably types too, right?

Comment (barthc) [reply packages/netlify-cms-core/src/actions/config.js]: Yes, `types` as well.

**Comment by ahlgorithm** [comment]
@ianhomer friendly ping. This PR fixes the only blocking issue for me adopting Netlify CMS for an internal project. :)

**Comment by ianhomer** [comment]
@ahlgorithm it looks like https://github.com/netlify/netlify-cms/pull/2576 is going address some of what is covered in this PR.  When that merges I'll take a look at this again to see if there's any further value in this PR.

**Comment by ianworth** [comment]
Closing as work is ongoing to create an improved solution with some API additions - see https://github.com/netlify/netlify-cms/issues/725#issuecomment-528495484

**Comment by barthc** [comment]
@ianworth  if you can extract the `metadata` fix from this PR, we can merge that.

**Pull request closed without merging**